ignite/src/main/java/org/hibernate/ogm/datastore/ignite/impl/IgniteDatastoreProvider.java

```diff
@@ -0,0 +1,195 @@
+import org.hibernate.ogm.query.spi.QueryParserService;
+import org.hibernate.service.spi.Configurable;
+import org.hibernate.service.spi.ServiceRegistryAwareService;
+import org.hibernate.service.spi.ServiceRegistryImplementor;
+import org.hibernate.service.spi.Startable;
+import org.hibernate.service.spi.Stoppable;
+
+/**
+ * Provides access to a Ignite instance
+ *
+ * @author Dmitriy Kozlov
+ *
+ */
+public class IgniteDatastoreProvider extends BaseDatastoreProvider
+							implements Startable, Stoppable, ServiceRegistryAwareService, Configurable {
+
+	private static final long serialVersionUID = 2278253954737494852L;
+
+	private static final Log log = LoggerFactory.getLogger();
+
+	protected JtaPlatform jtaPlatform;
+	protected IgniteEx cacheManager;
+	protected IgniteProviderConfiguration config;
+
+	private String gridName;
+	/** true - if we run inside the server node (for distributed tasks) */
+	private boolean localNode = false;
+
+	public IgniteCache<String, BinaryObject> getEntityCache(String entityName) {
+		String entityCacheName = getKeyProvider().getEntityCache( entityName );
+		return getCache( entityCacheName, true );
+	}
+
+	public IgniteCache<String, BinaryObject> getEntityCache(EntityKeyMetadata keyMetaData) {
+		String entityCacheName = getKeyProvider().getEntityCache( keyMetaData );
+		return getCache( entityCacheName, true );
+	}
+
+	private <T> IgniteCache<String, T> getCache(String entityCacheName, boolean keepBinary) {
+		IgniteCache<String, T> cache = cacheManager.cache( entityCacheName );
+		if (cache == null) {
+			log.warn("Unknown cache '" + entityCacheName + "'. Creating new with default settings.");
+			CacheConfiguration<String, T> config = new CacheConfiguration<>();
+			config.setName( entityCacheName );
+			cache = cacheManager.getOrCreateCache( config );
+		}
+		if (keepBinary) {
+			cache = ((IgniteCacheProxy<String, T>) cache).keepBinary();
+		}
+		return cache;
+	}
+
+	public IgniteCache<String, BinaryObject> getAssociationCache(AssociationKeyMetadata keyMetaData) {
+		String entityCacheName = getKeyProvider().getAssociationCache( keyMetaData );
+		return getCache(entityCacheName, true);
+	}
+
+	public IgniteCache<String, Object> getIdSourceCache(IdSourceKeyMetadata keyMetaData) {
+		String idSourceCacheName = getKeyProvider().getIdSourceCache( keyMetaData );
+		return getCache(idSourceCacheName, false);
+	}
+
+	public BinaryObjectBuilder getBinaryObjectBuilder(String type) {
+		return cacheManager.binary().builder( type );
+	}
+
+	@Override
+	public void configure(Map map) {
+		config = new IgniteProviderConfiguration();
+		config.initialize( map );
+	}
+
+	@Override
+	public void stop() {
+		if (cacheManager != null && !localNode) {
+			Ignition.stop( cacheManager.name(), true );
+		}
+	}
+
+	private String createGridName() {
+		String result = null;
+		if (config.getUrl() != null) {
+			result = config.getUrl().getPath();
+			result = result.replaceAll( "[\\,\\\",:,\\*,\\/,\\\\]", "_" );
```

Comment:
Wouldn't you risk to have collisions between grid names? A comment or javadoc would be nice